builder/simplify.cc

```diff
@@ -1187,51 +1187,103 @@ class simplifier : public node_mutator {
   }
 
   void visit(const let* op) override {
-    interval_expr value_bounds;
-    expr value = mutate(op->value, &value_bounds);
+    std::vector<std::pair<symbol_id, expr>> lets;
+    lets.reserve(op->lets.size());
+
+    using sv_type = std::pair<scoped_value_in_symbol_map<interval_expr>, scoped_value_in_symbol_map<int>>;
+    std::vector<sv_type> scoped_values;
+    scoped_values.reserve(op->lets.size());
+
+    bool values_changed = false;
+    for (const auto& s : op->lets) {
+      interval_expr value_bounds;
+      lets.emplace_back(s.first, mutate(s.second, &value_bounds));
+      values_changed = values_changed || !lets.back().second.same_as(s.second);
+
+      auto vb = set_value_in_scope(expr_bounds, s.first, value_bounds);
+      auto rc = set_value_in_scope(references, s.first, 0);
+      scoped_values.emplace_back(std::move(vb), std::move(rc));
+    }
 
-    auto set_bounds = set_value_in_scope(expr_bounds, op->sym, value_bounds);
-    auto ref_count = set_value_in_scope(references, op->sym, 0);
     interval_expr body_bounds;
     expr body = mutate(op->body, &body_bounds);
 
-    int refs = *references[op->sym];
-    if (refs == 0) {
-      // This let is dead
+    // First, prune any dead lets
+    for (auto it = lets.begin(); it != lets.end();) {
+      int refs = *references[it->first];
+      if (refs == 0) {
+        it = lets.erase(it);
+        values_changed = true;
+      } else {
+        it++;
+      }
+    }
+
+    // For any lets that have a single ref, or have a const-or-var value, just inline them
```

Comment:
Why not do this in one loop? The `erase` logic is a bit tricky (not your fault, it's just the way STL `erase` functions work) so just doing it once would be nice.